README.md
# mutwo.zimmermann

[![Build Status](https://circleci.com/gh/mutwo-org/mutwo.zimmermann.svg?style=shield)](https://circleci.com/gh/mutwo-org/mutwo.zimmermann)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![PyPI version](https://badge.fury.io/py/mutwo.zimmermann.svg)](https://badge.fury.io/py/mutwo.zimmermann)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

New functionality for event based library [mutwo](https://github.com/mutwo-org/mutwo) related to [Levin Eric Zimmermann](levinericzimmermann.eu).

This extension implements:

- `mutwo.zimmermann_generators`

### Installation

mutwo.zimmermann is available on [pypi](https://pypi.org/project/mutwo.zimmermann/) and can be installed via pip:

```sh
pip3 install mutwo.zimmermann
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.4.0] - 2022-10-08
- Added title package


## [0.4.0] - 2022-08-14

### Changed
- package name from `mutwo.ext-zimmermann` to `mutwo.zimmermann`


## [0.3.0] - 2022-03-06

### Added
- `euclidean_interlocking` in `zimmermann_generators`


## [0.2.0] - 2022-03-06

### Added
- `add_unison` option in the `from_constraints` class method of `PitchBasedContextFreeGrammar`




tests/generators_tests.py
import datetime
import typing
import unittest

import treelib

from mutwo import common_generators
from mutwo import zimmermann_generators


class PitchBasedContextFreeGrammarTests(unittest.TestCase):
    def setUp(cls):
        cls.pitch_based_context_free_grammar = (
            zimmermann_generators.PitchBasedContextFreeGrammar.from_constraints(
                prime_number_to_maximum_exponent_dict={
                    3: 1,
                    5: 1,
                },
                maximum_cent_deviation=550,
            )
        )

    def _create_tree(
        self,
        pitch_based_context_free_grammar: zimmermann_generators.PitchBasedContextFreeGrammar,
        start: zimmermann_generators.JustIntonationPitchNonTerminal,
        *leaf_data: tuple[
            typing.Union[
                zimmermann_generators.JustIntonationPitchNonTerminal,
                zimmermann_generators.JustIntonationPitchTerminal,
            ],
            ...,
        ],
    ) -> treelib.Tree:
        expected_tree = treelib.Tree()
        start_data = (start,)
        root = expected_tree.create_node(
            tag=pitch_based_context_free_grammar._data_to_tag(start_data),
            data=start_data,
        )
        for ld in leaf_data:
            expected_tree.create_node(
                tag=pitch_based_context_free_grammar._data_to_tag(ld),
                data=ld,
                parent=root,
            )
        return expected_tree

    def _compare_tree(self, expected_tree: treelib.Tree, real_tree: treelib.Tree):
        for expected_leaf, real_leaf in zip(expected_tree.leaves(), real_tree.leaves()):
            self.assertEqual(expected_leaf.data, real_leaf.data)
            self.assertEqual(expected_leaf.tag, real_leaf.tag)

    def test_from_constraints(self):
        expected_rule_tuple = (
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("4/5"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("3/4"),
                    zimmermann_generators.JustIntonationPitchTerminal("16/15"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("4/5"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("16/15"),
                    zimmermann_generators.JustIntonationPitchTerminal("3/4"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("5/4"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("4/3"),
                    zimmermann_generators.JustIntonationPitchTerminal("15/16"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("5/4"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("15/16"),
                    zimmermann_generators.JustIntonationPitchTerminal("4/3"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("4/3"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("5/4"),
                    zimmermann_generators.JustIntonationPitchTerminal("16/15"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("4/3"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("16/15"),
                    zimmermann_generators.JustIntonationPitchTerminal("5/4"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("3/4"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("4/5"),
                    zimmermann_generators.JustIntonationPitchTerminal("15/16"),
                ),
            ),
            common_generators.ContextFreeGrammarRule(
                zimmermann_generators.JustIntonationPitchNonTerminal("3/4"),
                (
                    zimmermann_generators.JustIntonationPitchTerminal("15/16"),
                    zimmermann_generators.JustIntonationPitchTerminal("4/5"),
                ),
            ),
        )
        for expected_rule, real_rule in zip(
            expected_rule_tuple,
            self.pitch_based_context_free_grammar.context_free_grammar_rule_tuple,
        ):
            self.assertEqual(expected_rule.left_side, real_rule.left_side)
            self.assertEqual(expected_rule.right_side, real_rule.right_side)

    def test_from_constraints_with_unison(self):
        pitch_based_context_free_grammar = (
            zimmermann_generators.PitchBasedContextFreeGrammar.from_constraints(
                prime_number_to_maximum_exponent_dict={},
                add_unison=True,
                maximum_cent_deviation=1200,
                allowed_octave_sequence=(-1, 0, 1),
            )
        )
        self.assertEqual(
            pitch_based_context_free_grammar.context_free_grammar_rule_tuple,
            (
                common_generators.ContextFreeGrammarRule(
                    zimmermann_generators.JustIntonationPitchNonTerminal("1/1"),
                    (
                        zimmermann_generators.JustIntonationPitchNonTerminal("1/2"),
                        zimmermann_generators.JustIntonationPitchNonTerminal("2/1"),
                    ),
                ),
                common_generators.ContextFreeGrammarRule(
                    zimmermann_generators.JustIntonationPitchNonTerminal("1/1"),
                    (
                        zimmermann_generators.JustIntonationPitchNonTerminal("2/1"),
                        zimmermann_generators.JustIntonationPitchNonTerminal("1/2"),
                    ),
                ),
            ),
        )
        self.assertEqual(
            pitch_based_context_free_grammar.non_terminal_tuple,
            (
                zimmermann_generators.JustIntonationPitchNonTerminal("1/2"),
                zimmermann_generators.JustIntonationPitchNonTerminal("1/1"),
                zimmermann_generators.JustIntonationPitchNonTerminal("2/1"),
            ),
        )
        self.assertEqual(
            pitch_based_context_free_grammar.terminal_tuple,
            tuple([]),
        )

    def test_resolve(self):
        start = zimmermann_generators.JustIntonationPitchNonTerminal("3/4")
        resolution = self.pitch_based_context_free_grammar.resolve(start, limit=1)
        expected_tree = self._create_tree(
            self.pitch_based_context_free_grammar,
            start,
            (
                zimmermann_generators.JustIntonationPitchTerminal("4/5"),
                zimmermann_generators.JustIntonationPitchTerminal("15/16"),
            ),
            (
                zimmermann_generators.JustIntonationPitchTerminal("15/16"),
                zimmermann_generators.JustIntonationPitchTerminal("4/5"),
            ),
        )
        self._compare_tree(expected_tree, resolution)

    def test_resolve_no_movement(self):
        pitch_based_context_free_grammar = (
            zimmermann_generators.PitchBasedContextFreeGrammar.from_constraints(
                prime_number_to_maximum_exponent_dict={
                    3: 1,
                },
                maximum_cent_deviation=500,
                add_unison=True,
            )
        )
        start = zimmermann_generators.JustIntonationPitchNonTerminal("1/1")
        resolution = pitch_based_context_free_grammar.resolve(start, limit=1)
        expected_tree = self._create_tree(
            pitch_based_context_free_grammar,
            start,
            (
                zimmermann_generators.JustIntonationPitchTerminal("3/4"),
                zimmermann_generators.JustIntonationPitchTerminal("4/3"),
            ),
            (
                zimmermann_generators.JustIntonationPitchTerminal("4/3"),
                zimmermann_generators.JustIntonationPitchTerminal("3/4"),
            ),
        )
        self._compare_tree(expected_tree, resolution)


class EuclideanInterlockingTest(unittest.TestCase):
    def test_euclidean_interlocking(self):
        sequence0, sequence1 = [0, 0, 0], [1, 1]
        self.assertEqual(
            zimmermann_generators.euclidean_interlocking(sequence0, sequence1),
            (0, 1, 0, 0, 1),
        )

    def test_empty_euclidean_interlocking(self):
        self.assertEqual(
            zimmermann_generators.euclidean_interlocking(),
            tuple([]),
        )

    def test_single_euclidean_interlocking(self):
        self.assertEqual(
            zimmermann_generators.euclidean_interlocking([1, 2, 3]),
            (1, 2, 3),
        )

    def test_euclidean_interlocking_with_empty_element(self):
        self.assertEqual(
            zimmermann_generators.euclidean_interlocking([1, 2, 3], [], []),
            (1, 2, 3),
        )


class TitleTest(unittest.TestCase):
    def test_golden_number(self):
        self.assertEqual(
            zimmermann_generators.golden_number(datetime.datetime(2022, 1, 1)), 9
        )
        self.assertEqual(
            zimmermann_generators.golden_number(datetime.datetime(2021, 1, 1)), 8
        )
        self.assertEqual(
            zimmermann_generators.golden_number(datetime.datetime(1, 1, 1)), 2
        )

    def test_get_title(self):
        self.assertEqual(
            zimmermann_generators.get_title(datetime.datetime(2022, 1, 1)),
            f"9.{zimmermann_generators.constants.GOLDEN_NUMBER_TO_COMPOSITION_COUNTER_DICT[9] + 1}",
        )


if __name__ == "__main__":
    unittest.main()


mutwo/zimmermann_generators/titles.py
import datetime
import typing

from mutwo import zimmermann_generators

__all__ = (
    "golden_number",
    "get_title",
)


def golden_number(date_time: typing.Optional[datetime.datetime] = None) -> int:
    """Get index in `metonic cycle <https://en.wikipedia.org/wiki/Metonic_cycle>`_.

    :param date_time: The current time. If set to ``None`` it will use current time.
    :type date_time: typing.Optional[datetime.datetime]
    :return: Number between 1 to 19. This is the current index of the metonic cycle.

    See `here <https://en.wikipedia.org/wiki/Golden_number_(time)>`_ for more
    information regarding golden numbers.
    """

    if date_time is None:
        date_time = datetime.datetime.now()

    year = date_time.year
    return (year % 19) + 1


def get_title(date_time: typing.Optional[datetime.datetime] = None) -> str:
    """Find title of current composition"""

    current_golden_number = golden_number(date_time)
    composition_counter = (
        zimmermann_generators.constants.GOLDEN_NUMBER_TO_COMPOSITION_COUNTER_DICT[
            current_golden_number
        ]
    ) + 1
    return f"{current_golden_number}.{composition_counter}"


mutwo/zimmermann_generators/zimmermann.py
"""Algorithms which are related to the mutwo author L.E. Zimmermann"""

from __future__ import annotations
import itertools
import typing

import primesieve
import treelib

from mutwo import common_generators
from mutwo import core_utilities
from mutwo import music_parameters

__all__ = (
    "PitchBasedContextFreeGrammar",
    "JustIntonationPitchTerminal",
    "JustIntonationPitchNonTerminal",
)


class JustIntonationPitchNonTerminal(
    music_parameters.JustIntonationPitch, common_generators.NonTerminal
):
    pass


class JustIntonationPitchTerminal(
    music_parameters.JustIntonationPitch, common_generators.Terminal
):
    pass


class PitchBasedContextFreeGrammar(common_generators.ContextFreeGrammar):
    """Create `JustIntonationPitch` sequences from single intervals.

    This class is a simple :class:`mutwo.common_generators.ContextFreeGrammar`
    with alternative constructor methods and an additional filter function
    which avoids adding undesired parts to the resulting :class:`treelib.Tree`.
    The implementation is based on the text "Bewegungen im unendlichen Tonraum"
    (2018) from L.E. Zimmermann.
    """

    @staticmethod
    def _maximum_exponent_to_allowed_exponent_tuple(
        maximum_exponent: int,
    ) -> tuple[int, ...]:
        return tuple(range(1, maximum_exponent + 1)) + tuple(
            range(-maximum_exponent, 0)
        )

    @staticmethod
    def _add_pitch_to_terminal_list_or_non_terminal_list(
        minimal_barlow_harmonicity_non_terminal: float,
        minimal_barlow_harmonicity_terminal: float,
        prime_number_tuple: tuple[int, ...],
        exponent_tuple: tuple[int, ...],
        ascending_prime_tuple: tuple[int, ...],
        non_terminal_list: list[JustIntonationPitchNonTerminal],
        terminal_list: list[JustIntonationPitchTerminal],
        octave: int,
        maximum_cent_deviation: float,
    ):
        complete_exponent_list = [
            exponent_tuple[prime_number_tuple.index(prime)]
            if prime in prime_number_tuple
            else 0
            for prime in ascending_prime_tuple
        ]
        just_intonation_pitch = music_parameters.JustIntonationPitch(
            complete_exponent_list
        ).register(octave)
        if abs(just_intonation_pitch.cents) <= maximum_cent_deviation:
            harmonicity_simplified_barlow = (
                just_intonation_pitch.harmonicity_simplified_barlow
            )
            if harmonicity_simplified_barlow >= minimal_barlow_harmonicity_non_terminal:
                non_terminal_list.append(
                    JustIntonationPitchNonTerminal(just_intonation_pitch.exponent_tuple)
                )
            elif harmonicity_simplified_barlow >= minimal_barlow_harmonicity_terminal:
                terminal_list.append(
                    JustIntonationPitchTerminal(just_intonation_pitch.exponent_tuple)
                )

    @staticmethod
    def _get_terminal_tuple_and_non_terminal_tuple(
        minimal_barlow_harmonicity_non_terminal: float,
        minimal_barlow_harmonicity_terminal: float,
        prime_number_to_maximum_exponent_dict: dict[int, int],
        allowed_octave_sequence: typing.Sequence[int],
        maximum_cent_deviation: float,
    ) -> tuple[
        tuple[JustIntonationPitchNonTerminal, ...],
        tuple[JustIntonationPitchTerminal, ...],
    ]:
        terminal_list = []
        non_terminal_list = []

        prime_number_tuple = tuple(prime_number_to_maximum_exponent_dict.keys())
        prime_number_to_allowed_exponent_tuple_dict = {
            prime_number: PitchBasedContextFreeGrammar._maximum_exponent_to_allowed_exponent_tuple(
                maximum_exponent
            )
            for prime_number, maximum_exponent in prime_number_to_maximum_exponent_dict.items()
        }
        # All occuring primes until the highest prime
        ascending_prime_tuple = tuple(primesieve.primes(max(prime_number_tuple)))

        for n_prime_numbers in range(1, len(prime_number_to_maximum_exponent_dict) + 1):
            for combined_prime_number_tuple in itertools.combinations(
                prime_number_tuple, n_prime_numbers
            ):
                for exponent_tuple in itertools.product(
                    *[
                        prime_number_to_allowed_exponent_tuple_dict[prime_number]
                        for prime_number in combined_prime_number_tuple
                    ]
                ):
                    for octave in allowed_octave_sequence:
                        PitchBasedContextFreeGrammar._add_pitch_to_terminal_list_or_non_terminal_list(
                            minimal_barlow_harmonicity_non_terminal,
                            minimal_barlow_harmonicity_terminal,
                            combined_prime_number_tuple,
                            exponent_tuple,
                            ascending_prime_tuple,
                            non_terminal_list,
                            terminal_list,
                            octave,
                            maximum_cent_deviation,
                        )
        return tuple(non_terminal_list), tuple(terminal_list)

    @classmethod
    def from_constraints(
        cls,
        minimal_barlow_harmonicity_non_terminal: float = 0.1,
        minimal_barlow_harmonicity_terminal: float = 0.05,
        prime_number_to_maximum_exponent_dict: dict[int, int] = {
            3: 2,
            5: 1,
            7: 1,
            11: 1,
        },
        allowed_octave_sequence: typing.Sequence[int] = (-1, 0),
        maximum_cent_deviation: float = 500,
    ) -> PitchBasedContextFreeGrammar:
        """Create rules based on various constraints.

        :param minimal_barlow_harmonicity:
        :param prime_number_sequence:
        :param allowed_octave_sequence:
        :param maximum_cent_deviation:
        """

        assert (
            minimal_barlow_harmonicity_non_terminal
            > minimal_barlow_harmonicity_terminal
        )

        (
            non_terminal_tuple,
            terminal_tuple,
        ) = cls._get_terminal_tuple_and_non_terminal_tuple(
            minimal_barlow_harmonicity_non_terminal,
            minimal_barlow_harmonicity_terminal,
            prime_number_to_maximum_exponent_dict,
            allowed_octave_sequence,
            maximum_cent_deviation,
        )

        element_tuple = non_terminal_tuple + terminal_tuple
        rule_list = []
        for element0, element1 in itertools.combinations(element_tuple, 2):
            summed = element0 + element1
            if summed in non_terminal_tuple:
                non_terminal = non_terminal_tuple[non_terminal_tuple.index(summed)]
                rule_list.extend(
                    (
                        common_generators.ContextFreeGrammarRule(
                            non_terminal, (element0, element1)
                        ),
                        common_generators.ContextFreeGrammarRule(
                            non_terminal, (element1, element0)
                        ),
                    )
                )
        return cls(rule_list)

    def _data_to_tag(
        self,
        data: tuple[
            typing.Union[JustIntonationPitchTerminal, JustIntonationPitchNonTerminal],
            ...,
        ],
    ) -> str:
        return "({})".format(
            " ".join([f"{pitch.numerator}/{pitch.denominator}" for pitch in data])
        )

    def _add_node(
        self,
        tree: treelib.Tree,
        data: tuple[
            typing.Union[JustIntonationPitchTerminal, JustIntonationPitchNonTerminal],
            ...,
        ],
        parent: typing.Optional[treelib.Node] = None,
    ):
        pitch_accumulation = tuple(
            core_utilities.accumulate_from_n(
                data, music_parameters.JustIntonationPitch("1/1")
            )
        )
        if len(core_utilities.uniqify_sequence(pitch_accumulation)) == len(
            pitch_accumulation
        ):
            super()._add_node(tree, data, parent)


mutwo/zimmermann_generators/extended_chomsky.py
"""Algorithms which are related to the mutwo author L.E. Zimmermann"""

from __future__ import annotations
import itertools
import typing

import treelib
from sympy import primerange

from mutwo import common_generators
from mutwo import core_utilities
from mutwo import music_parameters

__all__ = (
    "PitchBasedContextFreeGrammar",
    "JustIntonationPitchTerminal",
    "JustIntonationPitchNonTerminal",
)


class JustIntonationPitchNonTerminal(
    music_parameters.JustIntonationPitch, common_generators.NonTerminal
):
    pass


class JustIntonationPitchTerminal(
    music_parameters.JustIntonationPitch, common_generators.Terminal
):
    pass


class PitchBasedContextFreeGrammar(common_generators.ContextFreeGrammar):
    """Create `JustIntonationPitch` sequences from single intervals.

    This class is a simple :class:`mutwo.common_generators.ContextFreeGrammar`
    with alternative constructor methods and an additional filter function
    which avoids adding undesired parts to the resulting :class:`treelib.Tree`.
    The implementation is based on the text "Bewegungen im unendlichen Tonraum"
    (2018) from L.E. Zimmermann.

    **Example**:

    >>> from mutwo import zimmermann_generators
    >>> pitch_based_context_free_grammar = (
    >>>     zimmermann_generators.PitchBasedContextFreeGrammar.from_constraints(
    >>>         prime_number_to_maximum_exponent_dict={3: 1, 5: 1},
    >>>         maximum_cent_deviation=550,
    >>>     )
    >>> )
    >>> resolution = pitch_based_context_free_grammar.resolve(
    >>>     zimmermann_generators.JustIntonationPitchNonTerminal("5/4"),
    >>>     limit=3
    >>> )
    >>> resolution.show()
    (5/4)
    ├── (15/16 4/3)
    │   └── (15/16 5/4 16/15)
    │       └── (15/16 15/16 4/3 16/15)
    └── (4/3 15/16)
        └── (16/15 5/4 15/16)
            └── (16/15 4/3 15/16 15/16)
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._exponent_tuple_tuple_to_status = {}

    @staticmethod
    def _maximum_exponent_to_allowed_exponent_tuple(
        maximum_exponent: int,
    ) -> tuple[int, ...]:
        return tuple(range(1, maximum_exponent + 1)) + tuple(
            range(-maximum_exponent, 0)
        )

    @staticmethod
    def _add_pitch_to_terminal_list_or_non_terminal_list(
        minimal_barlow_harmonicity_non_terminal: float,
        minimal_barlow_harmonicity_terminal: float,
        prime_number_tuple: tuple[int, ...],
        exponent_tuple: tuple[int, ...],
        ascending_prime_tuple: tuple[int, ...],
        non_terminal_list: list[JustIntonationPitchNonTerminal],
        terminal_list: list[JustIntonationPitchTerminal],
        octave: int,
        maximum_cent_deviation: float,
    ):
        complete_exponent_list = [
            exponent_tuple[prime_number_tuple.index(prime)]
            if prime in prime_number_tuple
            else 0
            for prime in ascending_prime_tuple
        ]
        just_intonation_pitch = music_parameters.JustIntonationPitch(
            complete_exponent_list
        ).register(octave)
        if abs(just_intonation_pitch.interval) <= maximum_cent_deviation:
            harmonicity_simplified_barlow = (
                just_intonation_pitch.harmonicity_simplified_barlow
            )
            if harmonicity_simplified_barlow >= minimal_barlow_harmonicity_non_terminal:
                non_terminal_list.append(
                    JustIntonationPitchNonTerminal(just_intonation_pitch.exponent_tuple)
                )
            elif harmonicity_simplified_barlow >= minimal_barlow_harmonicity_terminal:
                terminal_list.append(
                    JustIntonationPitchTerminal(just_intonation_pitch.exponent_tuple)
                )

    @staticmethod
    def _get_terminal_tuple_and_non_terminal_tuple(
        minimal_barlow_harmonicity_non_terminal: float,
        minimal_barlow_harmonicity_terminal: float,
        prime_number_to_maximum_exponent_dict: dict[int, int],
        allowed_octave_sequence: typing.Sequence[int],
        maximum_cent_deviation: float,
        add_unison: bool,
    ) -> tuple[
        tuple[JustIntonationPitchNonTerminal, ...],
        tuple[JustIntonationPitchTerminal, ...],
    ]:
        terminal_list = []
        non_terminal_list = []

        prime_number_tuple = tuple(prime_number_to_maximum_exponent_dict.keys())
        prime_number_to_allowed_exponent_tuple_dict = {
            prime_number: PitchBasedContextFreeGrammar._maximum_exponent_to_allowed_exponent_tuple(
                maximum_exponent
            )
            for prime_number, maximum_exponent in prime_number_to_maximum_exponent_dict.items()
        }
        # All occuring primes until the highest prime
        try:
            ascending_prime_tuple = tuple(primerange(max(prime_number_tuple) + 1))
        # If there are no prime numbers we can simply use an empty tuple
        except ValueError:
            ascending_prime_tuple = tuple([])

        for n_prime_numbers in range(
            not add_unison, len(prime_number_to_maximum_exponent_dict) + 1
        ):
            for combined_prime_number_tuple in itertools.combinations(
                prime_number_tuple, n_prime_numbers
            ):
                for exponent_tuple in itertools.product(
                    *[
                        prime_number_to_allowed_exponent_tuple_dict[prime_number]
                        for prime_number in combined_prime_number_tuple
                    ]
                ):
                    for octave in allowed_octave_sequence:
                        PitchBasedContextFreeGrammar._add_pitch_to_terminal_list_or_non_terminal_list(
                            minimal_barlow_harmonicity_non_terminal,
                            minimal_barlow_harmonicity_terminal,
                            combined_prime_number_tuple,
                            exponent_tuple,
                            ascending_prime_tuple,
                            non_terminal_list,
                            terminal_list,
                            octave,
                            maximum_cent_deviation,
                        )
        return tuple(non_terminal_list), tuple(terminal_list)

    @classmethod
    def from_constraints(
        cls,
        minimal_barlow_harmonicity_non_terminal: float = 0.1,
        minimal_barlow_harmonicity_terminal: float = 0.05,
        prime_number_to_maximum_exponent_dict: dict[int, int] = {
            3: 2,
            5: 1,
            7: 1,
            11: 1,
        },
        allowed_octave_sequence: typing.Sequence[int] = (-1, 0),
        maximum_cent_deviation: float = 500,
        add_unison: bool = False,
    ) -> PitchBasedContextFreeGrammar:
        """Create rules based on various constraints.

        :param minimal_barlow_harmonicity:
        :param prime_number_sequence:
        :param allowed_octave_sequence:
        :param maximum_cent_deviation:
        :param add_unison:
        """

        assert (
            minimal_barlow_harmonicity_non_terminal
            > minimal_barlow_harmonicity_terminal
        )

        (
            non_terminal_tuple,
            terminal_tuple,
        ) = cls._get_terminal_tuple_and_non_terminal_tuple(
            minimal_barlow_harmonicity_non_terminal,
            minimal_barlow_harmonicity_terminal,
            prime_number_to_maximum_exponent_dict,
            allowed_octave_sequence,
            maximum_cent_deviation,
            add_unison,
        )

        element_iterator = filter(
            lambda terminal_or_non_terminal: terminal_or_non_terminal
            != music_parameters.JustIntonationPitch("1/1"),
            non_terminal_tuple + terminal_tuple,
        )
        rule_list = []
        for element0, element1 in itertools.combinations(element_iterator, 2):
            summed = element0 + element1
            if summed in non_terminal_tuple:
                non_terminal = non_terminal_tuple[non_terminal_tuple.index(summed)]
                rule_list.extend(
                    (
                        common_generators.ContextFreeGrammarRule(
                            non_terminal, (element0, element1)
                        ),
                        common_generators.ContextFreeGrammarRule(
                            non_terminal, (element1, element0)
                        ),
                    )
                )
        return cls(rule_list)

    def _data_to_tag(
        self,
        data: tuple[
            typing.Union[JustIntonationPitchTerminal, JustIntonationPitchNonTerminal],
            ...,
        ],
    ) -> str:
        return "({})".format(
            " ".join([f"{pitch.numerator}/{pitch.denominator}" for pitch in data])
        )

    def _add_node(
        self,
        tree: treelib.Tree,
        data: tuple[
            typing.Union[JustIntonationPitchTerminal, JustIntonationPitchNonTerminal],
            ...,
        ],
        parent: typing.Optional[treelib.Node] = None,
    ):
        exponent_tuple_tuple = tuple(
            just_intonation_pitch.exponent_tuple for just_intonation_pitch in data
        )
        if exponent_tuple_tuple in self._exponent_tuple_tuple_to_status:
            status = self._exponent_tuple_tuple_to_status[exponent_tuple_tuple]
        else:
            # We only add an element if the goal isn't already reached
            # in a step in between and if there aren't any pitch
            # repetitions.
            pitch_accumulation = tuple(
                core_utilities.accumulate_from_n(
                    data, music_parameters.JustIntonationPitch("1/1")
                )
            )

            # Special treatment for movement 1/1: here the first and the
            # last pitches are equal and we should skip one in order
            # to pass the test.
            if pitch_accumulation[0] == pitch_accumulation[-1]:
                pitch_accumulation = pitch_accumulation[1:]

            status = len(core_utilities.uniqify_sequence(pitch_accumulation)) == len(
                pitch_accumulation
            )
            self._exponent_tuple_tuple_to_status.update({exponent_tuple_tuple: status})

        if status:
            super()._add_node(tree, data, parent)
